### PR TITLE
move optimization in executors in distributed mode

### PR DESCRIPTION
**Pull request by rom1504**
reduce transfer and memory usage

### CODE REVIEW

File: autofaiss/indices/distributed.py [hunk 1]
```diff
@@ -287,16 +306,21 @@ def run(
...
         stage2_folder = temporary_indices_folder.rstrip("/") + "/stage-2"
-        next_stage_folder = _merge_to_n_indices(
-            spark_session=ss, n=100, src_folder=stage1_folder, dst_folder=stage2_folder
+        next_stage_folder, _ = _merge_to_n_indices(
+            spark_session=ss, n=100, src_folder=stage1_folder, dst_folder=stage2_folder, index_optimizer=None
         )
         if nb_indices_to_keep == 1:
-            merged_index = _merge_index(small_indices_folder=next_stage_folder, nb_batches=1)
+            merged_index, _ = _merge_index(small_indices_folder=next_stage_folder, nb_batches=1)
             fs.rm(temporary_indices_folder, recursive=True)
```
Comment (hitchhicker) [inline autofaiss/indices/distributed.py:316]: Now we can empty the temporary_indices_folder no matter what the nb_indices_to_keep is.

File: autofaiss/indices/distributed.py [hunk 2]
```diff
@@ -136,7 +140,8 @@ def _merge_index(
     end: Optional[int] = None,
     max_size_on_disk: str = "100GB",
     tmp_output_folder: Optional[str] = None,
-) -> faiss.Index:
+    index_optimizer: Callable = None,
+) -> Tuple[faiss.Index, Dict[str, str]]:
     """Merge all the indices in `small_indices_folder` into single one return the merged index."""
```
Comment (hitchhicker) [inline autofaiss/indices/distributed.py:145]: We are not only returning the merged index now, comment should be updated.

File: autofaiss/indices/distributed.py [hunk 3]
```diff
@@ -183,42 +188,55 @@ def _merge_from_local(merged: Optional[faiss.Index] = None) -> faiss.Index:
...
+    return merged_index, metric_infos
 
 
 def _get_file_system(path: str) -> fsspec.AbstractFileSystem:
     return fsspec.core.url_to_fs(path, use_listings_cache=False)[0]
 
 
-def _merge_to_n_indices(spark_session, n: int, src_folder: str, dst_folder: str):
+def _merge_to_n_indices(spark_session, n: int, src_folder: str, dst_folder: str, index_optimizer: Callable = None):
     """Merge all the indices from src_folder into n indices, and return the folder for the next stage"""
```
Comment (hitchhicker) [inline autofaiss/indices/distributed.py:208]: We should update the comment for the return thing.

**Review by hitchhicker — COMMENTED** [review]
Well done, it looks good, left some comments.

Have you run it successefully ?

File: autofaiss/external/quantize.py [hunk 4]
```diff
@@ -270,8 +269,26 @@ def write_ids_df_to_parquet(ids: pd.DataFrame, batch_id: int):
                 logger.debug(f"Writing id DataFrame to file {output_file}")
                 ids.to_parquet(f)
 
+        def index_optimizer(index, index_suffix):
+            cur_index_path = index_path + index_suffix
+            cur_index_infos_path = index_infos_path + index_suffix
+            if any(re.findall(r"OPQ\d+_\d+,IVF\d+_HNSW\d+,PQ\d+", index_key)):
+                set_search_hyperparameters(index, f"nprobe={64},efSearch={128},ht={2048}", use_gpu)
```
[the remark this answers is not in the document]
Comment (rom1504) [reply autofaiss/external/quantize.py:276]: indeed but that's exactly what was done before
first here https://github.com/criteo/autofaiss/blob/fc7ba51d1f34d7bab5481971af36ae2c04acdb1b/autofaiss/external/build.py#L219
then there https://github.com/criteo/autofaiss/blob/fc7ba51d1f34d7bab5481971af36ae2c04acdb1b/autofaiss/external/optimize.py#L426

I believe it's done on purpose so the index parameters have an initial value that is reasonable

File: autofaiss/indices/distributed.py [hunk 5]
```diff
@@ -183,42 +188,55 @@ def _merge_from_local(merged: Optional[faiss.Index] = None) -> faiss.Index:
...
     fs = _get_file_system(src_folder)
     nb_indices_on_src_folder = len(fs.ls(src_folder, detail=False))
-    if nb_indices_on_src_folder <= n:
```
[the remark this answers is not in the document]
Comment (rom1504) [reply autofaiss/indices/distributed.py:200]: indeed, added

File: autofaiss/indices/distributed.py [hunk 6]
```diff
@@ -183,42 +188,55 @@ def _merge_from_local(merged: Optional[faiss.Index] = None) -> faiss.Index:
...
-    )
+        return metrics
+
+    metrics_rdd = rdd.map(merge)
+    metrics = list(metrics_rdd.collect())
+    if index_optimizer is not None:
+        metrics_dict = {metric_info["index_path"]: metric_info for metric_info in metrics}
+    else:
+        metrics_dict = None
     fs = _get_file_system(src_folder)
```
[the remark this answers is not in the document]
Comment (rom1504) [reply autofaiss/indices/distributed.py]: removed

File: autofaiss/indices/distributed.py (hunk 1 above)
Comment (rom1504) [reply autofaiss/indices/distributed.py:324]: yes, done

File: autofaiss/indices/distributed.py (hunk 2 above)
Comment (rom1504) [reply autofaiss/indices/distributed.py]: done

File: autofaiss/indices/distributed.py (hunk 3 above)
Comment (rom1504) [reply autofaiss/indices/distributed.py]: done

File: autofaiss/external/quantize.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hitchhicker) [reply autofaiss/external/quantize.py:276]: Okay, no problem then.

**Comment by hitchhicker** [comment]
LGTM now, let's merge it when your test on larger data succeeds.

**Comment by rom1504** [comment]
seems to work end to end now, but need to be careful about availability of both disk and ram for merging
doing https://github.com/criteo/autofaiss/issues/53#issuecomment-1060120010 would reduce the need for ram (but I don't really like that means having 2 files for 1 index, finding a solution to this problem could be nice)

**Comment by rom1504** [comment]
ok I confirm this last step works for a PQ128 index of 5B items

rerunning end to end with PQ64 to make sure everything is good

**Review by hitchhicker — APPROVED** [review]
LGTM

**Comment by hitchhicker** [comment]
> ok I confirm this last step works for a PQ128 index of 5B items
> rerunning end to end with PQ64 to make sure everything is good

Well done ! Let's merge it, I am optimist that it'll wokr for PQ64 as well.

**Comment by rom1504** [comment]
> rerunning end to end with PQ64 to make sure everything is good

it worked indeed!
```
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
22/03/08 03:02:57 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable                                               
2022-03-08 03:02:59,225 [INFO]: Using 16 omp threads (processes), consider increasing --nb_cores if you have more
2022-03-08 03:02:59,225 [INFO]: Launching the whole pipeline 03/08/2022, 03:02:59   
2022-03-08 03:02:59,225 [INFO]: Reading total number of vectors and dimension 03/08/2022, 03:02:59
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5854/5854 [03:08<00:00, 30.98it/s]
2022-03-08 03:06:12,253 [INFO]: There are 5550336490 embeddings of dim 768
2022-03-08 03:06:12,253 [INFO]: >>> Finished "Reading total number of vectors and dimension" in 193.0281 secs                                                                                
2022-03-08 03:06:12,253 [INFO]:         Compute estimated construction time of the index 03/08/2022, 03:06:12
2022-03-08 03:06:12,253 [INFO]:                 -> Train: 16.7 minutes                                                                                                                       
2022-03-08 03:06:12,253 [INFO]:                 -> Add: 13.2 hours                        
2022-03-08 03:06:12,253 [INFO]:                 Total: 13.5 hours                                                                                                                            
2022-03-08 03:06:12,253 [INFO]:         >>> Finished "Compute estimated construction time of the index" in 0.0001 secs
2022-03-08 03:06:12,253 [INFO]:         Checking that your have enough memory available to create the index 03/08/2022, 03:06:12                                                             
2022-03-08 03:06:12,254 [INFO]: 18.6GB of memory will be needed to build the index (more might be used if you have more)
2022-03-08 03:06:12,254 [INFO]:         >>> Finished "Checking that your have enough memory available to create the index" in 0.0006 secs                                                    
2022-03-08 03:06:12,254 [INFO]:         Selecting most promising index types given data characteristics 03/08/2022, 03:06:12
2022-03-08 03:06:12,254 [INFO]:         >>> Finished "Selecting most promising index types given data characteristics" in 0.0001 secs                                                        
2022-03-08 03:06:12,254 [INFO]:         Creating the index 03/08/2022, 03:06:12            
2022-03-08 03:06:12,254 [INFO]:                 -> Instanciate the index OPQ64_448,IVF131072_HNSW32,PQ64x8 03/08/2022, 03:06:12                                                              
2022-03-08 03:06:12,255 [INFO]:                 >>> Finished "-> Instanciate the index OPQ64_448,IVF131072_HNSW32,PQ64x8" in 0.0012 secs
2022-03-08 03:06:12,256 [INFO]: The index size will be approximately 372.6GB                                                                                                                 
2022-03-08 03:06:12,256 [INFO]:                 -> Extract training vectors 03/08/2022, 03:06:12    
2022-03-08 03:06:12,256 [INFO]: Will use 4063232 vectors to train the index, that will use 18.6GB of memory
 99%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊ | 125/126 [00:34<00:00,  3.65it/s]
2022-03-08 03:06:46,535 [INFO]:                 >>> Finished "-> Extract training vectors" in 34.2798 secs         
2022-03-08 03:06:46,536 [INFO]:                 -> Training the index with 4063232 vectors of dim 768 03/08/2022, 03:06:46

WARNING clustering 4063232 points to 131072 centroids: please provide at least 5111808 training points
2022-03-08 03:35:37,028 [INFO]:                 >>> Finished "-> Training the index with 4063232 vectors of dim 768" in 1730.4922 secs
2022-03-08 03:35:37,050 [INFO]: The memory available for adding the vectors is 14.7GB(total available - used by the index)
2022-03-08 03:35:37,050 [INFO]: Will be using at most 1GB of ram for adding
2022-03-08 03:35:37,050 [INFO]:                 -> Adding the vectors to the index 03/08/2022, 03:35:37
2022-03-08 03:35:37,050 [INFO]: Using a batch size of 325520 (memory overhead 953.7MB)
2022-03-08 03:35:39,451 [INFO]:                 -> Adding indices 03/08/2022, 03:35:39
2022-03-08 11:00:40,880 [INFO]:                 >>> Finished "-> Adding indices" in 26701.4284 secs
2022-03-08 11:00:40,880 [INFO]:                 -> Merging indices 03/08/2022, 11:00:40
2022-03-08 11:37:06,632 [INFO]:                 >>> Finished "-> Merging indices" in 2185.7524 secs
2022-03-08 11:37:06,632 [INFO]:                 >>> Finished "-> Adding the vectors to the index" in 28889.5822 secs
2022-03-08 11:37:06,632 [INFO]:         >>> Finished "Creating the index" in 30654.3782 secs
2022-03-08 11:37:06,632 [INFO]: >>> Finished "Launching the whole pipeline" in 30847.4073 secs 
```

result 55 pieces like this:
```
{"index_key": "OPQ64_448,IVF131072_HNSW32,PQ64x8", "index_param": "nprobe=125,efSearch=250,ht=2048", "index_path": "s3://laion-us-east-1/indices/vit-l-14/imagePQ64/knn.index00", "size in bytes": 7473867536, "avg_search_speed_ms": 10.37896910613505, "99p_search_speed_ms": 21.391539416799784, "reconstruction error %": 42.14935004711151, "nb vectors": 100006060, "vectors dimension": 768, "compression ratio": 41.10570796715282}
```

**Pull request merged**